internal/flavors/publisher_test.go

```diff
@@ -208,3 +210,140 @@ func sleepOnCycleEnd(t *testing.T, size int, eventCount int, interval time.Durat
+		"single batch": {
+			interval:                            2 * time.Second,
+			threshold:                           10,
+			incomeEventBatchesOf:                []int{5},
+			expectedPublishedBatchesIDs:         [][]int{{1, 2, 3, 4, 5}},
+			expectedBufferCapacityOnEachPublish: []int{15},
+			expectedBufferCapacityEnd:           15,
+		},
+		"single batch increase capacity": {
+			interval:                            2 * time.Second,
+			threshold:                           5,
+			incomeEventBatchesOf:                []int{10},
+			expectedPublishedBatchesIDs:         [][]int{{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}},
+			expectedBufferCapacityOnEachPublish: []int{15},
+			expectedBufferCapacityEnd:           15,
+		},
+		"two batches under threshold": {
+			interval:                            2 * time.Second,
+			threshold:                           10,
+			incomeEventBatchesOf:                []int{5, 5},
+			expectedPublishedBatchesIDs:         [][]int{{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}},
+			expectedBufferCapacityOnEachPublish: []int{15},
+			expectedBufferCapacityEnd:           15,
+		},
+		"two batches at threshold": {
+			interval:                            2 * time.Second,
+			threshold:                           5,
+			incomeEventBatchesOf:                []int{3, 3, 6},
+			expectedPublishedBatchesIDs:         [][]int{{1, 2, 3, 4, 5, 6}, {7, 8, 9, 10, 11, 12}},
+			expectedBufferCapacityOnEachPublish: []int{7, 7},
+			expectedBufferCapacityEnd:           7,
+		},
+		"single batch over threshold": {
+			interval:                            2 * time.Second,
+			threshold:                           3,
+			incomeEventBatchesOf:                []int{5},
+			expectedPublishedBatchesIDs:         [][]int{{1, 2, 3, 4, 5}},
+			expectedBufferCapacityOnEachPublish: []int{8},
+			expectedBufferCapacityEnd:           8, // 4 * 2
+		},
+		"single batch over threshold break capacity limit": {
+			interval:                            2 * time.Second,
+			threshold:                           3,
+			incomeEventBatchesOf:                []int{15},
+			expectedPublishedBatchesIDs:         [][]int{{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15}},
+			expectedBufferCapacityOnEachPublish: []int{15},
+			expectedBufferCapacityEnd:           4,
+		},
+		"multiple batches break capacity limit": {
+			interval:                            2 * time.Second,
+			threshold:                           5,
+			incomeEventBatchesOf:                []int{2, 2, 20},
+			expectedPublishedBatchesIDs:         [][]int{{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24}},
+			expectedBufferCapacityOnEachPublish: []int{27},
+			expectedBufferCapacityEnd:           7,
+		},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			log := testhelper.NewLogger(t)
+			defer goleak.VerifyNone(t, goleak.IgnoreCurrent())
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			var publisher *Publisher
```

Comment:
```suggestion
			client := newMockClient(t)
			publisher := NewPublisher(log, time.Hour, tc.threshold, client)
```

IMO a bit cleaner

Reply:
I believe it's a matter of pov, I prefer having the mock created and initialized together.

Reply:
hmm, my preference is because the client mock is dereferencing a nil pointer. Ofc it's in a callback and it is not dangerous in any way currently, but still feels "wrong".

Up to you.